### PR TITLE
Use request.nonce when generating hybrid id token

### PR DESCRIPTION
**Pull request by tevansuk**
Like with the implicit grant, we need to override `add_id_token` to pass the nonce from the current request to `GrantBase.add_id_token` in order for the ID token to have the correct nonce.

Add test that the nonce is in ID token from hybrid OIDC flow.

Fixes: #746

### CODE REVIEW

**Comment by tevansuk** [comment]
I see two ways of fixing this. The first is what is in the PR, which follows how this is done in OIDC auth code flow and OIDC implicit flow, to override `add_id_token` and inject the `nonce`.

The second is to do away with all these overrides (and `RequestValidator.get_authorization_code_nonce`, and have `GrantTypeBase.add_id_token` to just look at `request.nonce`. For auth code flow, this would involve a change in behaviour for OIDC providers to require  `RequestValidator.validate_code` to set `request.nonce` when validating the code. This would be a backwards incompatible change however, and so I went with the first option.

File: oauthlib/openid/connect/core/grant_types/hybrid.py [hunk 1]
```diff
@@ -35,6 +35,9 @@ def __init__(self, request_validator=None, **kwargs):
         self.register_code_modifier(self.add_id_token)
         self.register_token_modifier(self.add_id_token)
 
+    def add_id_token(self, token, token_handler, request):
+        return super().add_id_token(token, token_handler, request, nonce=request.nonce)
```
Comment (auvipy) [inline oauthlib/openid/connect/core/grant_types/hybrid.py:39]: is this nonce related to this? https://stackoverflow.com/questions/3690738/need-help-understanding-nonce

Comment (tevansuk) [reply oauthlib/openid/connect/core/grant_types/hybrid.py:39]: Yep, or better description is [in the spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest). Its used to prevent replay attacks, the client generates a random nonce and stores it in a session, and then compares the nonce in the response with the nonce in the session. If it is missing either in the session or the response, or does not match, the response is invalid.

Its a required parameter for the implicit flow, a sometimes required parameter for the hybrid flow and an optional parameter for authorization code flow.

Handling it, however, is not optional - if a nonce is presented in the authorization endpoint, it must be in the ID token or a client should reject the response as invalid.

**Comment by JonathanHuot** [comment]
Hi @tevansuk , good catch ! However note that nonce is optional for "code token" in hybrid mode, as highlighted in https://github.com/oauthlib/oauthlib/blob/d54965b86ce4ede956db70baff0b3d5e9182a007/oauthlib/openid/connect/core/grant_types/base.py#L115-L123 ; however your code seems anyway using the right approach (using request.nonce will work even without value).

**Review by JonathanHuot — COMMENTED** [review]
LGTM

**Comment by tevansuk** [comment]
Thanks for the review!

I'm not sure that text is correct by the way. The spec says that it is an optional authentication request parameter in certain flows, and required in other flows, but it *can* be specified in all of them, and if it is used, the server *must* include it in the ID token, and clients *must* verify it  - [section 2](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) (ID Token)

> If present in the ID Token, Clients MUST verify that the nonce Claim Value is equal to the value of the nonce parameter sent in the Authentication Request. If present in the Authentication Request, Authorization Servers MUST include a nonce Claim in the ID Token with the Claim Value being the nonce value sent in the Authentication Request. Authorization Servers SHOULD perform no other processing on nonce values used.

Then in [section 3.3.2.11](https://openid.net/specs/openid-connect-core-1_0.html#HybridIDToken) (Hybrid Flow - ID Token):

> nonce
>     Use of the nonce Claim is REQUIRED for this flow.

For "code token" of course, the id token is retrieved by the auth code flow grant handler, which handles it the same as a standard auth code, so it did already work correctly for that case!

**Review by auvipy — APPROVED** [review]
